bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/server/enable/HasLanguageServerPropertyTester.java

```diff
@@ -26,9 +38,49 @@ public HasLanguageServerPropertyTester() {
 	@Override
 	public boolean test(Object receiver, String property, Object[] args, Object expectedValue) {
 		if (cLanguageServerProvider != null) {
-			return cLanguageServerProvider.isEnabledFor(receiver);
+			if (receiver instanceof URI) {
+				// called from the language server enabler for LSP4E:
+				var uri = (URI) receiver;
+				var fileHandle = getFileHandle(uri);
+				if(fileHandle != null && fileHandle instanceof IResource) {
+					var resource = (IResource) fileHandle;
+					if (resource != null && resource.getProject() != null) {					
+						return cLanguageServerProvider.isEnabledFor(resource.getProject());
+					}
+				}
+				// when resource == null it's an external file: Check if the file is already opened, if not check the active editor:
+				return LspUtils.isFileOpenedInLspEditor(uri);
+			} else if (receiver instanceof ITranslationUnit) {
+				// called to enable the LS based CSymbolsContentProvider: 
+				return Optional.of((ITranslationUnit) receiver)
+						 .map(ITranslationUnit::getCProject)
+						 .map(ICProject::getProject)
+						 .map(cLanguageServerProvider::isEnabledFor)
+						 .orElse(Boolean.FALSE);
+			} else if (receiver instanceof DocumentSymbolWithFile) {
+				// called to enable the LS based CSymbolsContentProvider:
+				return true;
+			} else if (receiver instanceof IProject) {
+				return cLanguageServerProvider.isEnabledFor((IProject) receiver);
+			}
+			return false;
+		}
+		return false;
+	}
+	
+	private IFile getFileHandle(URI uri) {
+		if (uri == null) {
+			return null;
+		}
+		if (FILE_SCHEME.equals(uri.getScheme())) {
+			IFile[] files = LspPlugin.getDefault().getWorkspace().getRoot().findFilesForLocationURI(uri);
+			if (files.length > 0) {
+				return files[0];
+			}
+			return null;
+		} else {
+			return Adapters.adapt(uri.toString(), IFile.class, true);
```

Comment:
I'm not aware of a use case where this could work.

Reply:
Do you mean that the adapt will fail if the scheme from uri is not "file"?

Reply:
This call tries to adapt a Sting value to an instance of `IFile` 
This possible only in theory, if someone registered such an adapter. Have you ever seen non-null value returned by this line?

Reply:
> Have you ever seen non-null value returned by this line?
Yes, it works at least for uri with file scheme. I haven't tested other schemes. It uses the `queryAdapterManager` in  `org.eclipse.core.runtime.Adapters.adapt(Object, Class<T>, boolean)`

Reply:
I haven't check which adapter will be used

Reply:
> Yes, it works at least for uri with file scheme

Cool! But if it works, do we need to do anything else, i.e. do we need to ask workspace root for locations and so on?

Reply:
> Cool! But if it works, do we need to do anything else, i.e. do we need to ask workspace root for locations and so on?

I just tested it again, and the path of the `IFile`returned by `Adapters.adapt(uri.toString(), IFile.class, true);` is not correct: 
/LSPEclipseUtilsTestfile/TR/workspaces/runtime-cdt-development/Cmk1/Cmk1.cpp
it should be:
/Cmk1/Cmk1.cpp
Cmk11 is my project.

Reply:
Interesting. What `IProject` is returned by such a file?

Reply:
> Interesting. What IProject is returned by such a file?

P/LSPEclipseUtilsTestfile

Reply:
Dear @ghentschke could you please remove [org.eclipse.lsp4e.test](https://github.com/eclipse/lsp4e/tree/master/org.eclipse.lsp4e.test) bundle from your .launch configuration and try again.

It seems that what you see is produced by this [adapter factory](https://github.com/eclipse/lsp4e/blob/9f06f405409541d2015438a7e153ef674b30e85f/org.eclipse.lsp4e.test/src/org/eclipse/lsp4e/test/UriToResourceAdapterFactory.java#L44)